File: qc-core/src/main/java/com/eager/questioncloud/core/domain/payment/QuestionPayment.java
package com.eager.questioncloud.core.domain.payment;

import com.eager.questioncloud.core.domain.coupon.Coupon;
import com.eager.questioncloud.core.domain.coupon.CouponType;
import com.eager.questioncloud.core.domain.question.Question;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import lombok.Builder;
import lombok.Getter;

@Getter
public class QuestionPayment {
    private Long id;
    private String paymentId;
    private List<QuestionPaymentOrder> orders;
    private Long userId;
    private Long userCouponId;
    private int amount;
    private QuestionPaymentStatus status;
    private LocalDateTime createdAt;

    @Builder
    public QuestionPayment(Long id, String paymentId, List<QuestionPaymentOrder> orders, Long userId, Long userCouponId, int amount,
        QuestionPaymentStatus status, LocalDateTime createdAt) {
        this.id = id;
        this.paymentId = paymentId;
        this.orders = orders;
        this.userId = userId;
        this.userCouponId = userCouponId;
        this.amount = amount;
        this.status = status;
        this.createdAt = createdAt;
    }

    public static QuestionPayment create(Long userId, Long userCouponId, List<Question> questions) {
        String paymentId = UUID.randomUUID().toString();
        int originalAmount = calcOriginalAmount(questions);
        List<QuestionPaymentOrder> orders = QuestionPaymentOrder.createOrders(paymentId, questions);

        return QuestionPayment.builder()
            .paymentId(paymentId)
            .orders(orders)
            .userId(userId)
            .userCouponId(userCouponId)
            .amount(originalAmount)
            .status(QuestionPaymentStatus.SUCCESS)
            .createdAt(LocalDateTime.now())
            .build();
    }

    private static int calcOriginalAmount(List<Question> questions) {
        return questions
            .stream()
            .mapToInt(question -> question.getQuestionContent().getPrice())
            .sum();
    }

    public Boolean isUsingCoupon() {
        return userCouponId != null;
    }

    public void useCoupon(Coupon coupon) {
        if (coupon.getCouponType().equals(CouponType.Fixed)) {
            amount = Math.max(amount - coupon.getValue(), 0);
        }
        if (coupon.getCouponType().equals(CouponType.Percent)) {
            int discountAmount = (amount * (coupon.getValue() / 100));
            amount = amount - discountAmount;
        }
    }

    public void fail() {
        this.status = QuestionPaymentStatus.FAIL;
    }
}

File: qc-payment/src/main/kotlin/com/eager/questioncloud/payment/domain/QuestionPayment.kt
package com.eager.questioncloud.payment.domain

import com.eager.questioncloud.payment.enums.QuestionPaymentStatus
import java.time.LocalDateTime

class QuestionPayment(
    var order: QuestionOrder,
    var userId: Long,
    var questionPaymentCoupon: QuestionPaymentCoupon?,
    var amount: Int,
    var status: QuestionPaymentStatus = QuestionPaymentStatus.SUCCESS,
    var createdAt: LocalDateTime = LocalDateTime.now(),
) {
    fun isUsedCoupon(): Boolean {
        return questionPaymentCoupon != null
    }

    fun applyCoupon() {
        amount = questionPaymentCoupon!!.calcDiscount(amount)
    }

    fun fail() {
        this.status = QuestionPaymentStatus.FAIL
    }

    companion object {
        fun create(
            userId: Long,
            questionPaymentCoupon: QuestionPaymentCoupon?,
            order: QuestionOrder
        ): QuestionPayment {
            return QuestionPayment(
                order = order,
                userId = userId,
                questionPaymentCoupon = questionPaymentCoupon,
                amount = order.calcAmount(),
            )
        }
    }
}
